File: src/Java/496.NextGreaterElementI.java

import java.awt.List;
import java.util.LinkedList;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.HashSet;

class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        int m = nums1.length;
        int n = nums2.length;
        int[] r = new int[m];
        HashMap<Integer, Integer> store = new HashMap<>();
        LinkedList<Integer> s = new LinkedList<>();
        for (int max : nums2) {
            while (s.size() > 0 && s.getLast() < max) {
                store.put(s.getLast(), max);
                s.removeLast();
            }
            s.add(max);
        }
        for (int i = 0; i < m; i++) {
            r[i] = store.getOrDefault(nums1[i], -1);
        }
        return r;
    }
}

File: src/Python/401-500/496.NextGreaterElementI.py

class Solution:
    def nextGreaterElement(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: List[int]
        """
        m = len(nums1)
        n = len(nums2)
        r = [-1] * m
        for j in range(m):
            num = nums1[j]
            index = nums2.index(num)
            for i in range(index + 1, n):
                if nums2[i] > num:
                    r[j] = nums2[i]
                    break
        return r

if __name__ == '__main__':
    solution = Solution()
    print(solution.nextGreaterElement([4, 1, 2], [1, 3, 4, 2]))
    print(solution.nextGreaterElement([2, 4], [1, 2, 3, 4]))
    print(solution.nextGreaterElement([1, 2], [3, 2, 1]))
else:
    pass
